Rewrite recipe chain calculator

Meegooo opened the pull request:
Rewrote recipe chain calculator. The main difference is that it now uses graph traversal to calculate things. Fixes https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/17148

Notable features:
- Support for "ore dictionary" (but really for multiple ingredients per slot)
- Ability to request negative amount of items to indicate that you already have some (disabled by default).
- Proper handling of recipes with multiple outputs and recursive recipes

There are some differences in how it looks and feels. Here are new behaviors.
1. Modifying item counts
    - Ctrl+scroll works only on recipe outputs and increases requested amount of only that item by 1. Fluids are increased by whatever one cell contains, or, if there is no cell, by whatever container holds the max amount. I don't like how the method looks (`StackInfo::getFluidCellSize`), but could not find a better approach.
    - Ctrl+Shift+Scroll works on any ingredient of the recipe and increases amount of requested items by whatever one craft gives you. This is notably different from how it was before, which was "increase number of requested items by one craft worth for the very last recipe in the chain". I decided to not reimplement this because of potential ambiguity, and because you can simply scroll on the recipe that you actually want.
    - Alt adds a multiplier that is equal to number under NEI item list on the bottom right.

2. Visual
   - The number of requested items is shown in the top right corner as `+123`/`-123`. This is not required to be a multiple of whatever the recipe output is. Previously it would just change number of items individually. This is also shown as `X*64 + Y` on the tooltip
   - Bottom right shows however many items will be consumed/produced by however many craft iterations were calculated.
   - With shift held, top rights shows number of craft operations in the top right for the whole chain as `x123`. Bottom right numbers show chain inputs and chain remainders/outputs. Background highlights behave the same way.
   - Shift highlights group that your cursor is next to. You don't need to put you mouse over an itemstack. Tooltips were annoying.

3. Behavioral and implementation details.
   - Item without ingredients in the crafting chain will sum requested ingredients from other recipes.
   - Pressing Shift+A on an output stack in a NEI recipe pins all output stacks, not just the one you hovered over.  I want opinions on that one. Changed it to be consistent with doing the same thing over plus sign. And in general, I don't like how you can only remove ingredients and outputs from pinned recipe, but adding them back is practically impossible. I have no idea how to fix it though. 
   - Amount of requested items is completely separate from pinned stack size that is shown. This means that toggling crafting mode will result in zeroed out item stack sizes. This also means that pinning item counts along with ingredients (Ctrl+Shift+A vs Shift+A) has no actual difference for this functionality.
   - If there are multiple recipes that give the same output, first code collects any remainders from all nodes, then, if it needs any more, it chooses first recipe it encounters. Currently, this item in any other recipes is highlighted in red, although that might not be needed anymore.
   - Graph traversal continues for an ingredient only if that ingredient is pinned as both input, and output in other recipe.
   - Recipes are still serialized and stored as `BookmarkRecipeId`. I extract actual recipes at runtime.
   - Mapping from original ItemStack to calculated ItemStack is done via indexes. Previously it was done as a mapping from ItemStack to ItemStack.

Some screenshots (some are not from GTNH, so textures might be different):

<details>
  <summary>Old screenshots</summary>
  
![image](https://github.com/user-attachments/assets/44a8e285-8478-41af-ad07-44dfc57e077b)
![image](https://github.com/user-attachments/assets/41e2f0dc-7367-4b23-832a-a713253437e9)
![image](https://github.com/user-attachments/assets/e2d9ba47-7fd0-4dc0-8e1b-960bb78378c8)
![image](https://github.com/user-attachments/assets/f613e748-cc28-42c8-82a6-e76f24a1d148)
  
</details>

![image](https://github.com/user-attachments/assets/ec8c318b-fd51-4d48-b7c9-1619eeef7ff7)
![image](https://github.com/user-attachments/assets/75cb5c3d-a77c-44ff-b9d1-d5b09a492431)
![image](https://github.com/user-attachments/assets/603eedca-4db6-4fd9-941a-af7205100c4b)
![image](https://github.com/user-attachments/assets/7e68661e-5133-42fc-a783-e9c9ba701023)
![image](https://github.com/user-attachments/assets/537eee8a-bc86-4270-825c-876dc41e07cc)
![image](https://github.com/user-attachments/assets/cdbd3da4-9022-4858-b69e-91236d7043a1)

slprime commented:
`ctrl+shift+scroll` don't scroll full recipe, only item under mouse. (work only ctrl+scroll)
![image](https://github.com/user-attachments/assets/5bec60aa-5adc-43d4-a233-f128921a3e8b)

why used second recipe? `has ingredients pinned` what you meen?
![image](https://github.com/user-attachments/assets/4edf9b08-2fd6-425b-8ac3-3d5d3b4ead95)

Recipe from first group scrolled too
![Minecraft1 7 102024-09-1521-13-08-ezgif com-video-to-gif-converter](https://github.com/user-attachments/assets/283376d7-dc9d-4875-988b-796b0b27a792)

> Alt adds a multiplier that is equal to number under NEI item list on the bottom right.

Don't work correctly. The value must be a multiple of 64 (in my example) but have 65
| before | after |
|-|-|
| ![image](https://github.com/user-attachments/assets/2b666fbd-c72b-4c2d-8de1-2226ca9b4999) | ![image](https://github.com/user-attachments/assets/4358ae9f-830a-4801-b3f4-789f5709b028) |

I'm not sure there is a need for a negative main recipe. I often use shift+ctrl+alt+scroll to reset the recipe.
![image](https://github.com/user-attachments/assets/809ac0d1-dc8c-4caf-8405-9ac8618f3799)

Shift in fluid show how many it is by detail (29kL=29000=....) But if use negative value i don't know how many fluids I've specified. For items same problem
| Positive | Negative |
|-|-|
| ![image](https://github.com/user-attachments/assets/67b2d969-4840-4167-a105-987093847546) | ![image](https://github.com/user-attachments/assets/12622272-f05d-4f08-97be-ac364e30fec2) |

The ingredient "Magnesiumchloride Dust" is a catalyst in this recipe group, it should not be in the ingredients and remainders. The display of circuits is also questionable
| Recipe | Tooltip |
|-|-|
| ![image](https://github.com/user-attachments/assets/a339eeda-5cad-47ab-b6e7-5f46409d2913) | ![image](https://github.com/user-attachments/assets/596c6b3f-d452-460d-a363-1c292abf0427) |

Same problem. Results have 72 chips but should have 64
| Recipe | Tooltip |
|-|-|
| ![image](https://github.com/user-attachments/assets/a986cadc-a5fe-4b2b-b6be-16cbe8e6aa86) | ![image](https://github.com/user-attachments/assets/8b651616-5ff5-451e-a0bf-7b3006ad337c) |

2 different recipes form "Cryogenic Freezer" and "Vacuum Freezer", but only one of them scrolls.
![image](https://github.com/user-attachments/assets/0ece8e12-7234-4839-bc89-441c9489df87)

Recipe Group
1. oil -> "Sulfuric Light Fuel"
2. "Sulfuric Light Fuel" -> "Light Fuel"
3. "Light Fuel" -> "Lightly Steam-Cracked Light Fuel"
4. "Lightly Steam-Cracked Light Fuel" -> Ethylene

Chain brocken because fluid in cell (it will work fine before)
| Recipe | shift |
|-|-|
| ![image](https://github.com/user-attachments/assets/44764ec6-0353-4e10-8741-d9042d96953d) | ![image](https://github.com/user-attachments/assets/e27d27cc-d31b-45a1-82f6-f7a05518c5ec) |

"Sulfuric Light Fuel" and "Light Fuel" highlight like it is in results/remainders, but it isn't
| Recipe | Tooltip |
|-|-|
| ![image](https://github.com/user-attachments/assets/76c90d80-8b89-4ed1-9f1e-c25d283bfbbd)| ![image](https://github.com/user-attachments/assets/86e078fb-8889-44e2-9d91-ba8bd1250ac9) |

Previously, the highlight was more informative. Using shift showed how many items were left
| Before | Now |
|-|-|
| ![image](https://github.com/user-attachments/assets/04f1d172-44c9-48fe-a6c7-be90d3f55bfa) | ![image](https://github.com/user-attachments/assets/9c4a1717-5443-4e04-b839-ce8382179a59) |

That's enough for now

Meegooo commented:
> ctrl+shift+scroll don't scroll full recipe, only item under mouse. (work only ctrl+scroll)

Here you have crafting chain mode disabled. Did it use to do that on shift+ctrl+scroll?

> why used second recipe?
> 2 different recipes form "Cryogenic Freezer" and "Vacuum Freezer", but only one of them scrolls.

It uses whatever recipe it encounters first. I should probably make it more consistent, because currently to the user it feels random because of hashmap. And maybe color conflicting recipe in gray or something. But functionally, I don't see a way to support multiple recipes that give the same output. If there is a recipe that you need pinned, but that you don't wanna use for some item, you can unpin output from that recipe.

>  has ingredients pinned what you meen?

If you unpin an ingredient from the recipe, then graph traversal won't continue, even if that recipe is pinned. For example, iron is not pinned in hopper recipe, so iron is not calculated. 
![image](https://github.com/user-attachments/assets/7d050020-b09d-47c7-8dc5-bc252f919216)
That filtering required extra (annoying) code. So i'm happy to remove that and always continue through for all ingredients.

> Recipe from first group scrolled too

Oops, forgot to filter by group id.

> Don't work correctly. The value must be a multiple of 64 (in my example) but have 65
> I'm not sure there is a need for a negative main recipe. I often use shift+ctrl+alt+scroll to reset the recipe.

Cause currently it adds/removes 64 or whatever multiple you have set. But now that I think about it, forcing it to be a multiple would be nicer. Noted. 
Adding shift would increase/decrease number of crafting operations, not reuqested items. So, for 6 slabs from 3 planks it would scroll to -320, 0, 320, 640, 960, etc.

> Chain brocken because fluid in cell (it will work fine before)

You mean it should treat cells as raw fluids? Makes sense, noted.

> The ingredient "Magnesiumchloride Dust" is a catalyst in this recipe group, it should not be in the ingredients and remainders. The display of circuits is also questionable
> Same problem. Results have 72 chips but should have 64

That is by design. It is a catalyst that is unused, but it is required for the craft. They were cancelling each other in first implementation, but after some playtesting I found out that it was not very helpful. In general, this tries to mimic the way AE2 autocrafting works. And that's how it would see it. 
As for circuits, not showing ingredients with stack size of 0 would solve it.

> "Sulfuric Light Fuel" and "Light Fuel" highlight like it is in results/remainders, but it isn't

That's actually Lightly Steam-Cracked fuel in the remainders. Ethylene requires 19000L and cracking recipe will produce 19200. That is a problem I had with the tooltip, because you often can't tell what ingredient/results/remainders are shown, because a lot of them look very similar 
Sulfuric light fuel is also in the remainders. You did not explicitly request it (you requested only Ethylene), but it was produced as a byproduct of converting sulfuric fuel to normal fuel. Now, if you Ctrl+scroll up on it twice, it will be considered something you requested and will move from remainders to results.

> Using shift showed how many items were left
You mean showing outputs+remainders as a stack size for outputs? Okay, previously I couldn't exactly figure out what that was.

> Shift in fluid show how many it is by detail (29kL=29000=....) But if use negative value i don't know how many fluids I've specified. For items same problem. 

Do you know where that is calculated? There is no reason it shouldn't work, other than explicitly checking for stack size to be greater than 0.

slprime commented:
> Here you have crafting chain mode disabled. Did it use to do that on shift+ctrl+scroll?

Previously, the algorithm worked the same whether the chain was enabled or not

> It uses whatever recipe it encounters first. I should probably make it more consistent, because currently to the user it feels random because of hashmap. And maybe color conflicting recipe in gray or something. But functionally, I don't see a way to support multiple recipes that give the same output. If there is a recipe that you need pinned, but that you don't wanna use for some item, you can unpin output from that recipe.

Here it would be correct to change the quantity on the recipe you are on. 
How will it behave if it is an intermediate recipe? will i be able to specify how many recipes should go on one path and how many on the other or will i have to create several groups?

> Adding shift would increase/decrease number of crafting operations, not reuqested items. So, for 6 slabs from 3 planks it would scroll to -320, 0, 320, 640, 960, etc.

How to reset group state? Why do you need a negative value in the final recipe?

> You mean it should treat cells as raw fluids? Makes sense, noted.

yes, it will work fine before

> That is by design. It is a catalyst that is unused, but it is required for the craft. They were cancelling each other in first implementation, but after some playtesting I found out that it was not very helpful. In general, this tries to mimic the way AE2 autocrafting works. And that's how it would see it.

This will mislead the player into crafting a catalyst in quantities he doesn't need. they are no different from a circuit in situations like this. The only thing you can do is to output it as an ingredient in the amount of one item. Otherwise the player will again have to filter out unnecessary things in tooltips/highlights

> That's actually Lightly Steam-Cracked fuel in the remainders. Ethylene requires 19000L and cracking recipe will produce 19200. That is a problem I had with the tooltip, because you often can't tell what ingredient/results/remainders are shown, because a lot of them look very similar

correctly: Ethylene requires 18800L and cracking recipe will produce 19000. 
Previously, by holding down the shift key, I could clearly understand what I would have left and at what stage, now I have to look for inputs and outputs for item/fluid and subtract one number from another.

> Do you know where that is calculated? There is no reason it shouldn't work, other than explicitly checking for stack size to be greater than 0.

`GuiContainerManager.itemCountDetails`

Meegooo commented:
Notes to self, to not forget anything

- [x] Make output conflict resolution predictable (always pick the topmost recipe)
- [x] Forbid changing request amount for conflicting recipe
- [ ] Color conflicting recipe in red and show tooltip with hint (only when shift is held, to reduce noise)
- [x] Stop changing request count with Ctrl+Shift+Scroll for same recipe in another group
- [ ] Make scrolling consistent between regular mode and crafting mode. 
  - Ctrl+Scroll - change requests/stack size by 1 (or once cell worth for fluids) for only selected item (for crafting mode, only output items)
  - Ctrl+Shift+Scroll - change requests/stack size by `meta.factor` for all items in selected recipe (for crafting mode, only output items, inputs will be calculated)
  - Alt - adds an extra modifier.
  - Final count sticks to a multiple of final multiplier (`meta.factor*alt_modifier` or `cell_size*alt_modifier` or `1*alt_modifier`, etc)
- [ ] Make negative requests a config option (disable by default)
- [ ] Add config option to remove unconsumed catalysts and common items from inputs and outputs+remainders (enabled by default).
- [ ] Add number of requested items to tooltip (to know exact amount when it gets compressed to K/M/G)

Meegooo commented:
This should be ready, but I haven't playtested much. For this reason, I'm gonna keep it in draft state for a bit.

There is one known old bug, which is unrelated to the changes. You can't pin duping recipes (Null catalyst for example). But I would rather fix it in a new PR.

PlayfulPiano commented:
> This should be ready, but I haven't playtested much. For this reason, I'm gonna keep it in draft state for a bit.
> 
> There is one known old bug, which is unrelated to the changes. You can't pin duping recipes (Null catalyst for example). But I would rather fix it in a new PR.

is this the same bug causing issues with like for example multiple recipes using the same item? like tin & copper dust both using 3 xp buckets and therefore overwrite each other in bookmarks

Meegooo commented:
> is this the same bug causing issues with like for example multiple recipes using the same item? like tin & copper dust both using 3 xp buckets and therefore overwrite each other in bookmarks

Which recipe is this exactly? There was a bug that caused multiple stacks of same item in output to override each other when pinning recipe (ctrl+shift+A on plus sign). That one is fixed. The one that prevents you from properly pinning a recipe that has same items in inputs and outputs is not fixed.

PS. Regarding this PR in general. My attempt to remove common items from inputs and outputs (not remainders, those are fine) caused another bug, that prevents you from requesting i.e. 64 MV circuits and 64 HV circuits in one chain (64 MV circuits you requested will be consumed by 32 HV circuits). I could remove common items as a postprocessing step, but I would rather not. 

This would mean that whatever you requested will be shown in outputs exactly. This means that outputs and remainders will show how many items you will have in your inventory at the end, not how many you will gain.

PlayfulPiano commented:
> > is this the same bug causing issues with like for example multiple recipes using the same item? like tin & copper dust both using 3 xp buckets and therefore overwrite each other in bookmarks
> 
> Which recipe is this exactly? There was a bug that caused multiple stacks of same item in output to override each other when pinning recipe instead of adding up like inputs (ctrl+shift+A on plus sign). That one is fixed. The one that prevents you from properly pinning a recipe that has same items in inputs and outputs is not fixed.
> 
> PS. Regarding this PR in general. My attempt to remove common items from inputs and outputs (not remainders, those are fine) caused another bug, that prevents you from requesting i.e. 64 MV circuits and 64 HV circuits in one chain (64 MV circuits you requested will be consumed by 32 HV circuits). I could remove common items as a postprocessing step, but I would rather not.
> 
> This would mean that whatever you requested will be shown in outputs exactly. This means that outputs and remainders will show how many items you will have in your inventory at the end, not how many you will gain.

for example tin and copper dust both require 3 xp buckets in a crafting table, but if you try to bookmark both they'll overwrite each other.

Meegooo commented:
> for example tin and copper dust both require 3 xp buckets in a crafting table, but if you try to bookmark both they'll overwrite each other.

Ah, recipes override each other, not items, I see.  Currently recipe is identified by its inputs (shapeless) and crafting handler. That is not getting changed here. 
To fix this, recipe id should also include outputs (or shape information, but that needs a whitelist of shaped crafting handlers). That change is not trivial and needs to be done very carefully to not nuke any pinned recipes. That is to say, I'm not touching that here, but noted.

slprime commented:
### Added items from this gui (it added like pseudo recipe)
![image](https://github.com/user-attachments/assets/1c2aedd2-b9b6-44ad-a30a-127aea4acc9b)

### Highlight broken when chain on
| without chain | with chain |
|-|-|
| ![image](https://github.com/user-attachments/assets/ad7fd991-7a28-41a5-9ea6-b010e0ca854e) | ![image](https://github.com/user-attachments/assets/46e7a89b-ad21-4eb6-bb67-2eaaacd5cea2) |

### And it sometimes crashes when i hover to result with shift:
bookmarks state: 
![image](https://github.com/user-attachments/assets/cafe0f5d-ace7-41eb-a7b4-f45044f23503)

crash: https://mclo.gs/vQpnZBV

### It is also inconvenient that the count is reset when the chain is turned on
![image](https://github.com/user-attachments/assets/a4252d7c-1621-4f93-85e1-fb458efa812b)

### It is also inconvenient that all chain react to shift regardless of the cursor position

### Incorrect show chain remainders
| recipe | tooltip |
|-|-|
| ![image](https://github.com/user-attachments/assets/2ad2fc52-a118-4fe9-b74d-e0a1d9312929) | ![image](https://github.com/user-attachments/assets/13cd55a7-4b68-43d3-84cf-b4180d680d67) |

### If i added another recipe to chain, it incorrect calculate fluid. it use ingredient as result
| without | with |
|-|-|
| ![image](https://github.com/user-attachments/assets/c007a5df-85e2-4292-86e6-f0d039f745db) | ![image](https://github.com/user-attachments/assets/786394ce-bcd6-46aa-bf0b-d8edd6aeda33) |

Meegooo commented:
> ### Highlight broken when chain on
> ### Incorrect show chain remainders
> ### If i added another recipe to chain, it incorrect calculate fluid. it use ingredient as result
> ### Feedback from other user

All of those were caused by the bug I described above in this comment https://github.com/GTNewHorizons/NotEnoughItems/pull/530#issuecomment-2402858462. But now outputs can never be used as ingredients, and fixing that is not worth it in my opinion, especially for something that is this minor and arguably is not even a bug.

> ### ctrl+shift+a should be added only result under mouse, not all results. for all results exists `+`

Fixed

> ### Added items from this gui (it added like pseudo recipe)
> ...
> ### And it sometimes crashes when i hover to result with shift:

Couldn't reproduce adding the recipe, but I think it crashed because it couldn't find `craft-confirm` handler. I added a fallback that will use pinned ingredients instead of the real recipe

> ### It is also inconvenient that all chain react to shift regardless of the cursor position

I don't really understand this issue. It used to highlight the entire chain before this? The only difference is that now you don't need to hover over item, you can hover next to the group. And I don't see that as an issue.

> ### It is also inconvenient that the count is reset when the chain is turned on

Now, this is not trivial. Converting from crafting mode to regular mode is easy, but going backwards would require guessing however many items you need to request to best match counts of regular mode. I could request only roots of the chain, but that could reset requested amount of any intermediates (i.e. if you request both HV and EV circuits in one chain).

PS. I haven't had much free time for several weeks, so this is not playtested.

Meegooo commented:
Rebased and briefly tested the changes on 2.7.2 (with updated codechickencore and gtnhlib) and everything seems to work fine.
